Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt

```diff
@@ -0,0 +1,191 @@
+package de.rki.coronawarnapp.datadonation.analytics
+
+import de.rki.coronawarnapp.appconfig.AnalyticsConfig
+import de.rki.coronawarnapp.appconfig.AppConfigProvider
+import de.rki.coronawarnapp.appconfig.ConfigData
+import de.rki.coronawarnapp.appconfig.SafetyNetRequirements
+import de.rki.coronawarnapp.datadonation.analytics.modules.DonorModule
+import de.rki.coronawarnapp.datadonation.analytics.modules.exposureriskmetadata.ExposureRiskMetadataDonor
+import de.rki.coronawarnapp.datadonation.analytics.server.DataDonationAnalyticsServer
+import de.rki.coronawarnapp.datadonation.analytics.storage.AnalyticsSettings
+import de.rki.coronawarnapp.datadonation.analytics.storage.LastAnalyticsSubmissionLogger
+import de.rki.coronawarnapp.datadonation.safetynet.DeviceAttestation
+import de.rki.coronawarnapp.server.protocols.internal.ppdd.PpaData
+import de.rki.coronawarnapp.server.protocols.internal.ppdd.PpacAndroid
+import de.rki.coronawarnapp.storage.LocalData
+import io.mockk.MockKAnnotations
+import io.mockk.Runs
+import io.mockk.clearAllMocks
+import io.mockk.coEvery
+import io.mockk.coVerify
+import io.mockk.every
+import io.mockk.impl.annotations.MockK
+import io.mockk.just
+import io.mockk.mockkObject
+import io.mockk.spyk
+import org.joda.time.Days
+import org.joda.time.Instant
+import org.junit.jupiter.api.AfterEach
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import testhelpers.BaseTest
+import testhelpers.coroutines.runBlockingTest2
+import testhelpers.preferences.mockFlowPreference
+
+class AnalyticsTest : BaseTest() {
+    @MockK lateinit var dataDonationAnalyticsServer: DataDonationAnalyticsServer
+    @MockK lateinit var appConfigProvider: AppConfigProvider
+    @MockK lateinit var deviceAttestation: DeviceAttestation
+    @MockK lateinit var settings: AnalyticsSettings
+    @MockK lateinit var configData: ConfigData
+    @MockK lateinit var analyticsConfig: AnalyticsConfig
+    @MockK lateinit var exposureRiskMetadataDonor: ExposureRiskMetadataDonor
+    @MockK lateinit var lastAnalyticsSubmissionLogger: LastAnalyticsSubmissionLogger
+
+    @BeforeEach
+    fun setup() {
+        MockKAnnotations.init(this)
+
+        mockkObject(LocalData)
+
+        coEvery { appConfigProvider.getAppConfig() } returns configData
+        every { configData.analytics } returns analyticsConfig
+
+        coEvery { lastAnalyticsSubmissionLogger.storeAnalyticsData(any()) } just Runs
+    }
+
+    @AfterEach
+    fun tearDown() {
+        clearAllMocks()
+    }
+
+    private fun createDonorModules(): Set<DonorModule> = setOf(exposureRiskMetadataDonor)
+
+    private fun createInstance() = spyk(
```

Comment:
Why do we need  `spy` here?

Reply:
i was trying to use a spy to check for the return value of the internal check functions in Analytics i.e. `stopDueToProbabilityToSubmit` to not only check that submission was aborted as it is done now, but make sure that it is aborted for the right reason, maybe throwing a more specific exception in the stop functions and raising the exception handling one level further up in the analytics submission could also help with testability

Reply:
Maybe set up the `setup()` that the default state is equal to a successful submission, then in each test case (e.g. probability) change a single parameter and check that the process fails appropriately, maybe by checking that no submission happened.

Or if `stopDueToProbabilityToSubmit` causes an abort, then check that `stopDueToNoUserConsent` was called but `stopDueToLastSubmittedTimestamp`, maybe by verifying calls on the mocks used by those functions.

Reply:
I like it :)